### PR TITLE
Initial draft for a metrics, an API for exporters.

### PR DESCRIPTION
**Pull request by bogdandrutu**
This API can be seen as the api for the metrics/stats exporters, and it is not intended to replace the stats API.

The metrics API is intended to be used by the stats API as well as other metric systems (dropwizard, etc.) to send data using the census metric exporters framework.

Why do we need a different API for exporters:
* The current OpenCensus stats API is hard to be used for real GAUGE metrics (any pre-aggregated metric, e.g. machine-cpu-usage).
* The current OpenCensus stats exporter API is focus only on supporting OpenCensus stats, but the library needs to support different external sources (dropwizard, etc.) and supporting these sources is very hard with the current OpenCensus stats exporter API.

Here is an example of how to map ViewData into the Metrics:
```
MetricPoint
  MetricDescriptor
     Name <- ViewData.View.name
     Description <- ViewData.View.description
     Unit <- ViewData.View.Measure.unit
     Type <- Cumulative (except LastValue which is GAUGE)
  TimeSeries
     Labels <- {ViewData.rows, ViewData.View.columns}
  CumulativePoint
     Value <- ViewData.AggregationData.value (for DistributionValue the bucket_bounds are defined in the ViewData.View.DistributionAggregation.bucket_bounds)
     Start_time <- ViewData.start_time
     End_time <- ViewData.end_time
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -21,6 +21,7 @@ sourceSets {
     main {
         proto {
             // In addition to the default 'src/main/proto'
+            srcDir 'opencensus/proto/metrics'
```
Comment (songy23) [inline build.gradle]: Should this be under stats/metrics?

File: opencensus/proto/metrics/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2017, OpenCensus Authors
```
Comment (songy23) [inline opencensus/proto/metrics/BUILD.bazel]: nit: s/2017/2018

File: opencensus/proto/metrics/metrics.proto [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  MetricDescriptor desctriptor = 1;
+
+  // One or more timeseries for a single metric, where each timeseries has
+  // one or more points.
+  repeated TimeSeries timeseries = 2;
+}
+
+// Defines a metric type and its schema.
+message MetricDescriptor {
+  // The metric type, including its DNS name prefix. It must be unique.
```
Comment (songy23) [inline opencensus/proto/metrics/metrics.proto]: Should we mention the constraint (and possibly, [sanitization rules](https://github.com/census-instrumentation/opencensus-specs/issues/35)) on MetricDescriptor names?

File: opencensus/proto/metrics/metrics.proto [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  // The set of label values that uniquely identify this metric. Apply to all
+  // points.
+  repeated Label label = 1;
+
+  // The data points of this timeseries.
+  repeated Point points = 2;
+}
+
+message Label {
+  // The label name.
```
Comment (songy23) [inline opencensus/proto/metrics/metrics.proto]: Similarly, should we mention the constraints here?

File: opencensus/proto/metrics/metrics.proto [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+  // The label value.
+  string value = 2;
+}
+
+// One of the possible data points. The type of the point MUST match with the
+// type from the MetricDescriptor.
+message Point {
+  oneof point {
+    GaugePoint gauge_point = 1;
+    CumulativePoint points = 2;
```
Comment (songy23) [inline opencensus/proto/metrics/metrics.proto]: nit: s/points/cumulative_point

File: opencensus/proto/metrics/BUILD.bazel (hunk 2 above)
Comment (bogdandrutu) [reply opencensus/proto/metrics/BUILD.bazel]: Done.

File: build.gradle (hunk 1 above)
Comment (bogdandrutu) [reply build.gradle]: Done.

File: opencensus/proto/metrics/metrics.proto (hunk 3 above)
Comment (bogdandrutu) [reply opencensus/proto/metrics/metrics.proto]: Added a TODO for the moment.

File: opencensus/proto/metrics/metrics.proto (hunk 4 above)
Comment (bogdandrutu) [reply opencensus/proto/metrics/metrics.proto]: Similar, added a TODO for the moment.

File: opencensus/proto/metrics/metrics.proto (hunk 5 above)
Comment (bogdandrutu) [reply opencensus/proto/metrics/metrics.proto]: Done.

File: opencensus/proto/stats/metrics/metrics.proto [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+option java_outer_classname = "MetricsProto";
+
+// A collection of MetricPoint's.
+message MetricPointSet {
+  // Each MetricPoint has one or more timeseries for a single metric.
+  repeated MetricPoint metric_points = 1;
+}
+
+// Defines a MetricPoint which has one or more timeseries for a single metric.
+message MetricPoint {
```
Comment (rakyll) [inline opencensus/proto/stats/metrics/metrics.proto]: Shouldn't this be named MetricTimeSeries? MetricPoint makes it sound like it represents only one point.

Comment (g-easy) [reply opencensus/proto/stats/metrics/metrics.proto]: Why not just "Metric"?
I agree that "point" is confusing.

File: opencensus/proto/stats/metrics/metrics.proto [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+  // The description of the metric.
+  MetricDescriptor desctriptor = 1;
```
Comment (g-easy) [inline opencensus/proto/stats/metrics/metrics.proto]: trip -> rip

File: opencensus/proto/stats/metrics/metrics.proto [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+  }
+  Type type = 4;
+
+  // TODO: Consider to add LabelInfo{key, description} if needed.
+}
+
+// A collection of data points that describes the time-varying values
+// of a metric.
+message TimeSeries {
+  // The set of label values that uniquely identify this metric. Apply to all
```
Comment (g-easy) [inline opencensus/proto/stats/metrics/metrics.proto]: this metric -> this timeseries

File: opencensus/proto/stats/metrics/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+cc_proto_library(
+    name = "metrics_proto_cc",
+    deps = [":metrics_proto"],
+)
+
+java_proto_library(
+    name = "metrics_proto_java",
+    deps = [":metrics_proto"],
+)
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply opencensus/proto/stats/metrics/BUILD.bazel]: done.

File: opencensus/proto/stats/metrics/metrics.proto [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+package opencensus.proto.stats.metrics;
+
+import "google/protobuf/timestamp.proto";
+
+option go_package = "metricsproto";
+option java_multiple_files = true;
+option java_package = "io.opencensus.proto.metrics";
+option java_outer_classname = "MetricsProto";
+
+// A collection of MetricPoint's.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply opencensus/proto/stats/metrics/metrics.proto]: done.

File: opencensus/proto/stats/metrics/metrics.proto (hunk 6 above)
Comment (bogdandrutu) [reply opencensus/proto/stats/metrics/metrics.proto]: went with Metric/MetricSet. Let me know if it makes sense.

File: opencensus/proto/stats/metrics/metrics.proto (hunk 7 above)
Comment (bogdandrutu) [reply opencensus/proto/stats/metrics/metrics.proto]: done

File: opencensus/proto/stats/metrics/metrics.proto (hunk 8 above)
Comment (bogdandrutu) [reply opencensus/proto/stats/metrics/metrics.proto]: done.

File: opencensus/proto/stats/metrics/metrics.proto [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+  // TODO: Consider to add LabelInfo{key, description} if needed.
+}
+
+// A collection of data points that describes the time-varying values
+// of a metric.
+message TimeSeries {
+  // The set of label values that uniquely identify this metric. Apply to all
+  // points.
+  repeated Label label = 1;
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply opencensus/proto/stats/metrics/metrics.proto]: Done.

**Pull request merged**